src/operator/broadcast_reduce_op-inl.h

```diff
@@ -0,0 +1,189 @@
+  mshadow::Tensor<xpu, 1> in =
+      src.get_with_shape<xpu, 1, real_t>(mshadow::Shape1(src.shape_.Size()), s);
+  mshadow::VectorDot(out, in, in);
+  out = mshadow::expr::F<mxnet::op::mshadow_op::square_root>(out);
+}
+
+template<typename xpu, typename Reducer>
+void Reduce(const TBlob &src,
+            const EnvArguments& env,
+            TBlob *ret,
+            OpReqType req,
+            RunContext ctx) {
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  mshadow::Tensor<xpu, 1> out = ret->get<xpu, 1, real_t>(s);
+  mshadow::Tensor<xpu, 2> in =
+      src.get_with_shape<xpu, 2, real_t>(mshadow::Shape2(1, src.shape_.Size()), s);
+  out = mshadow::expr::reduce_except_dim<0, Reducer>(in);
+}
+
+template <typename xpu, typename Reducer>
+void ReduceMid(TBlob const& src,
+               const EnvArguments& env,
+               TBlob* ret,
+               OpReqType,
+               RunContext ctx) {
+  mshadow::Stream<xpu>* s = ctx.get_stream<xpu>();
+  mshadow::Tensor<xpu, 2> out = ret->get<xpu, 2, real_t>(s);
+  mshadow::Tensor<xpu, 3> in = src.get<xpu, 3, real_t>(s);
+  out = mshadow::expr::reduce_with_axis<Reducer, false>(in, 1);
+}
+
+inline TShape ReduceMidShape(const TShape& ishape,
+                             const EnvArguments& env)  {
+  CHECK_EQ(ishape.ndim(), 3) << "Input shape must be 3 dimensional.";
+  std::vector<mshadow::index_t> shape;
+  shape.push_back(ishape[0]);
+  shape.push_back(ishape[2]);
+  return TShape(shape.begin(), shape.end());
+}
+
+template<typename xpu, typename Reducer, bool get_mask>
+void ReduceChannel(const TBlob &src,
+                   const EnvArguments& env,
+                   TBlob *ret,
+                   OpReqType req,
+                   RunContext ctx) {
+  using namespace mxnet::op;
+  using namespace mshadow;
+  using namespace mshadow::expr;
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  Tensor<xpu, 2> out = ret->get_with_shape<xpu, 2, real_t>(
+    Shape2(src.shape_[0], src.Size()/src.shape_[0]/src.shape_[1]),
+    s);
+  Tensor<xpu, 3> in = src.get_with_shape<xpu, 3, real_t>(
+    Shape3(src.shape_[0], src.shape_[1], src.Size()/src.shape_[0]/src.shape_[1]),
+    s);
+  out = reduce_with_axis<Reducer, get_mask>(in, 1);
+}
+
+// return a shape of ReduceChannel output
+inline TShape ReduceChannelShape(const TShape& ishape,
+                                 const EnvArguments& env) {
+  std::vector<mshadow::index_t> shape;
+  shape.push_back(ishape[0]);
+  for (index_t i = 2; i < ishape.ndim(); ++i) {
+    shape.push_back(ishape[i]);
+  }
+  return TShape(shape.begin(), shape.end());
+}
+
+
+template<typename xpu>
+void Transpose(const TBlob &src,
+               const EnvArguments& env,
+               TBlob *ret,
+               OpReqType req,
+               RunContext ctx) {
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  mshadow::Tensor<xpu, 2> out = ret->FlatTo2D<xpu, real_t>(s);
+  mshadow::Tensor<xpu, 2> in = src.FlatTo2D<xpu, real_t>(s);
+  out = in.T();
```

Comment:
is this inplace safe? If not is there mechanism that guarentee src and ret are different?